Prepare cpuset for import into kubernets/utils

iancoolidge opened the pull request:
Preparation for cpuset to be moved into k/utils

Various API improvements from code review for library prep.

https://github.com/kubernetes/kubernetes/issues/112899

/kind cleanup

iancoolidge committed:
cpuset: Convert Union arguments to variadic

This allows Union to implement UnionAll easily.

iancoolidge committed:
cpuset: Remove *Int64 methods

These are rarely used and can be accommodated with a trivial helper.

iancoolidge committed:
cpuset: Remove 'MustParse' method

Removes exit/fatal from cpuset library.

Usage in podresources test was not necessary.

Library reference in cpu_manager_test was moved to a local function, and
converted to use e2e test framework error catching.

iancoolidge committed:
cpuset: Delete 'builder' methods

All usage of builder pattern is convertible to cpuset.New()
with the same or fewer lines of code.

Migrate Builder.Add to a private method of CPUSet, with a comment
that it is only intended for internal use to preserve immutable
propoerty of the exported interface.

This also removes 'require' library dependency, which avoids
non-standard library usage.

iancoolidge committed:
cpuset: Make 'ToSlice*' methods look like 'set' methods

In 'set', conversions to slice are done also, but with different names:

ToSliceNoSort() -> UnsortedList()
ToSlice() -> List()

Reimplement List() in terms of UnsortedList to save some duplication.

iancoolidge commented:
@thockin

iancoolidge commented:
/sig-node

pacoxu commented:
/labels tide/merge-method-squash
/ok-to-test
/release-note-none

iancoolidge commented:
/retest

klueska commented:
/assign
I'm not opposed to this in principle, but will need to take a closer look once the tree is open for 1.27 again.

bart0sh commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -162,20 +162,17 @@ func (s CPUSet) Difference(s2 CPUSet) CPUSet {
-	}
+// List returns a slice of integers that contains all elements from
+// this set. The list is sorted.
+func (s CPUSet) List() []int {
```
would it be more logical to call it SortedList? It's not obvious that List() result is sorted without knowing that there is an UnsortedList().

iancoolidge replied:
I am only naming it List() because this mirrors the 'set' API.

In 'set', conversions to slice are done also, but with different names:

ToSliceNoSort() -> UnsortedList()
ToSlice() -> List()

(see https://github.com/kubernetes/kubernetes/pull/113744/commits/bc95376e283ce179eb42bf4b2081f3d3d9b99038 for the commit that does this)

bart0sh commented:
/priority important-longterm
/triage accepted

thockin reviewed (commented):
AFAICT the only caller of FilterNot() is in cpuset.go - can we make that a private method?  Same with Filter() I think?

The only callers I see for NewBuilder() and Add() seem like they could be changed to populate a slice and then call NewCPUSet().

NewCPUSet()-> New() ?

Basically: play "what if I got rid of this" for everything exposed in the API, and see how small we can make it :)

iancoolidge committed:
cpuset: hide 'Filter' API

FilterNot is only used in this file, and is trivially converted to a
'filter' call site by inverting the predicate.

Filter is only used in this file, so don't export it.

iancoolidge committed:
cpuset: Rename 'NewCPUSet' to 'New'

iancoolidge commented:
> AFAICT the only caller of FilterNot() is in cpuset.go - can we make that a private method? Same with Filter() I think?
> 
> The only callers I see for NewBuilder() and Add() seem like they could be changed to populate a slice and then call NewCPUSet().
> 
> NewCPUSet()-> New() ?
> 
> Basically: play "what if I got rid of this" for everything exposed in the API, and see how small we can make it :)

Thanks Tim. Did all of above.

I made the builder a private function. I did not remove it entirely because:

removing the builder necesitates adding an 'Insert' method, which breaks the immutable property. I am fine removing this, but as far as I understand, it will be difficult to assert that we can safely remove the immutable property here, because it is hard to determine which callers/users are depending on that property (intentionally or not)

thockin commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -19,25 +19,22 @@ package cpuset
-func NewBuilder() *Builder {
-	return &Builder{
+// newBuilder returns a mutable CPUSet builder.
+func newBuilder() *builder {
```
I don't actually object to this, but it feels unnecessary - every place that uses it could just be:

```
s := New()
s.add(x)
s.add(y)
```

or even

```
s := New()
s.elems[x] = struct{}{}
```

It's only mutable from within this package, and having the builder doesn't make it much safer, and only slightly more obvious if someone was trying to mutate it when they shouldn't.

iancoolidge replied:
Good idea. Thanks Tim for the thorough reivew.

Done.

thockin reviewed (commented):
package cpuset could use a good package comment explaining that it's not just a simple set of ints because it can convert to/from the kernel's cpuset syntax.

Otherwise LGTM

iancoolidge committed:
cpuset: Add package comment

Describe use cases (node IDs, HT siblings, etc)

Call out novelty (Linux CPU list parse/dump)

Describe future work (relax immutable, refactor to use 'set')

iancoolidge commented:
> package cpuset could use a good package comment explaining that it's not just a simple set of ints because it can convert to/from the kernel's cpuset syntax.
> 
> Otherwise LGTM

Done also.

bart0sh commented:
/lgtm

iancoolidge commented:
Sorry for noise. Cleaned up code in test/e2e_node/cpu_manager_test.go a bit to use frameworks.ExpectNoError instead of dragging in klog and exit calls.

iancoolidge commented:
/retest

klueska commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -14,75 +14,48 @@ See the License for the specific language governing permissions and
-	return b.result
-}
-
 // CPUSet is a thread-safe, immutable set-like data structure for CPU IDs.
 type CPUSet struct {
 	elems map[int]struct{}
 }
 
```
It's still unclear to me why we don't just implement this in terms of a `sets.Set[int]` from the `apimachinery/pkg/util/sets` package. Most of the functions are already implemented (so would be a simple passthrough), and the one's that aren't can easily be implemented in terms of the others.

I had previously suggested this in https://github.com/kubernetes/utils/pull/267#discussion_r1028565778, but no one responded.

The only reason I can see *not* doing this is if that would prevent us from moving the implementation to `kubernetes/utils`.  I don't know what the requirements are there in terms of package imports.

/cc @thockin

iancoolidge replied:
Hi Kevin,

First, I agree that is a good goal (it is acknolwedged in the package comment, and has been discussed and not ignored, despite how your comment thread appears)

If you review my issue though, the problem I'm trying to solve is exactly exposing this in k/utils, so, depending on something *not* available in k/utils seems counter productive to the primary goal.

To me, a sensible strategy is:
1) keep current implementation of cpuset
2) move cpuset to k/utils
3) move set to k/utils also
4) reimplement cpuset in terms of set at that point

iancoolidge commented:
/retest

thockin commented on pkg/kubelet/cm/cpuset/cpuset.go:
```diff
@@ -14,75 +14,48 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// cpuset represnts a set of CPUs in a 'set' data structure.
```
Godoc wants `// Package cpuset`

s/represnts/represents

iancoolidge replied:
Done.

thockin reviewed (commented):
Just niggles in comments

thockin commented:
Thanks!

/lgtm
/approve
